File: scripts/mcp.lua
require "const"

local function create_start_points (self)
   local Distance = 10
   local Space = 5
   local Offset = math.modf (self.maxCycles / 4) + 1
   local Ident = dmz.matrix.new ()
   local HeadingPi = dmz.matrix.new (const.Up, dmz.math.Pi)
   local oddPos = { -(Space * Offset), 0, Distance }
   local evenPos = { -(Space * Offset) + (Space / 2), 0, -Distance }
   for ix = 1, self.maxCycles do
      local Odd = (1 == math.fmod (ix, 2))
      local point = {
         index = ix,
         pos = (Odd and oddPos or evenPos),
         ori = (Odd and Ident or HeadingPi),
         handle = dmz.object.create (const.StartPointType),
      }
      self.startPoints[ix] = point
      dmz.object.position (point.handle, nil, point.pos)
      dmz.object.orientation (point.handle, nil, point.ori)
      dmz.object.activate (point.handle)
      dmz.object.set_temporary (point.handle)
      if Odd then oddPos[1] = oddPos[1] + Space
      else evenPos[1] = evenPos[1] + Space
      end
   end
end

local function assign_points_to_cycles (self)
   self.playerCount = 0
   for object, cycle in pairs (self.cycleList) do
      if cycle.point then
         dmz.object.unlink_sub_links (cycle.point.handle, const.StartLinkHandle)
         cycle.point = nil
      elseif cycle.wait then
      end
   end
   local count = 1
   for object, cycle in pairs (self.cycleList) do
      if count <= self.maxCycles then
          cycle.point = self.startPoints[count]
          dmz.object.link (const.StartLinkHandle, cycle.point.handle, object)
          self.playerCount = self.playerCount + 1
      else
      end
      count = count + 1
   end
end

local function get_standby_count (self)
   local result = 0
   for object, cycle in pairs (self.cycleList) do
      if cycle.standby then result = result + 1 end
   end
   return result
end

local function get_dead_count (self)
   local result = 0
   for object, cycle in pairs (self.cycleList) do
      if cycle.dead then result = result + 1 end
   end
   return result
end

local function set_game_state (mcp, State)
   local newState = dmz.object.state (mcp)
   newState:unset (const.GameStateMask)
   newState = newState + State
   dmz.object.state (mcp, nil, newState)
end

local function update_time_slice (self, time)

   local State = dmz.object.state (self.mcp)

   if State then
      local CTime = dmz.time.frame_time ()
      if State:contains (const.GameWaiting) then
         if self.assignPoints then
            assign_points_to_cycles (self)
            self.assignPoints = false
         end
         if (get_standby_count (self) == self.playerCount) and
               (self.playerCount >= self.minCycles) then
self.log:error ("Move to Countdown 5")
            set_game_state (self.mcp, const.GameCountdown5)
            self.nextStateTime = CTime + 1
         end
      elseif State:contains (const.GameActive) then
         if get_dead_count (self) >= (self.playerCount - 1) then
            set_game_state (self.mcp, const.GameWaiting)
         end
      elseif State:contains (const.GameCountdown5) then
         if CTime >= self.nextStateTime then
self.log:error ("Move to Countdown 4")
            set_game_state (self.mcp, const.GameCountdown4)
            self.nextStateTime = CTime + 1
         end
      elseif State:contains (const.GameCountdown4) then
         if CTime >= self.nextStateTime then
self.log:error ("Move to Countdown 3")
            set_game_state (self.mcp, const.GameCountdown3)
            self.nextStateTime = CTime + 1
         end
      elseif State:contains (const.GameCountdown3) then
         if CTime >= self.nextStateTime then
self.log:error ("Move to Countdown 2")
            set_game_state (self.mcp, const.GameCountdown2)
            self.nextStateTime = CTime + 1
         end
      elseif State:contains (const.GameCountdown2) then
         if CTime >= self.nextStateTime then
self.log:error ("Move to Countdown 1")
            set_game_state (self.mcp, const.GameCountdown1)
            self.nextStateTime = CTime + 1
         end
      elseif State:contains (const.GameCountdown1) then
         if CTime >= self.nextStateTime then
self.log:error ("Move to Active")
            set_game_state (self.mcp, const.GameActive)
            self.nextStateTime = CTime + 1
         end
      else
         self.log.error ("Game in unknown state. Changing to a waiting state")
         set_game_state (self.mcp, const.GameWaiting)
      end
   end
end

local function create_object (self, Object, Type)
   if Type:is_of_type (const.CycleType) then
      self.cycleList[Object] = {}
      self.cycleCount = self.cycleCount + 1
      self.assignPoints = true
   end
end

local function destroy_object (self, Object)
   local cycle = self.cycleList[Object]
   if cycle then
      self.cycleList[Object] = nil
      self.cycleCount = self.cycleCount - 1
      if cycle.point then self.playerCount = self.playerCount - 1 end
      self.assignPoints = true
   end
end

local function update_object_state (self, Object, Attribute, State, PreviousState)
   local cycle = self.cycleList[Object]
   if cycle then
      if not PreviousState then PreviousState = const.EmptyState end
      if State:contains (const.Dead) and
            not PreviousState:contains (const.Dead) then
         cycle.dead = true
         cycle.standby = false
      elseif State:contains (const.Standby)
            and not PreviousState:contains (const.Standby) then
         cycle.dead = false
         cycle.standby = true
      else
         cycle.dead = false
         cycle.standby = false
      end
   end
end

local function start (self)
   self.handle = self.timeSlice:create (update_time_slice, self, self.name)
   local callbacks = {
      create_object = create_object,
      destroy_object = destroy_object,
      update_object_state = update_object_state,
   }
   self.objObs:register (nil, callbacks, self)
   self.mcp = dmz.object.create (const.MCPType)
   dmz.object.state (self.mcp, nil, const.GameWaiting)
   dmz.object.activate (self.mcp)
   dmz.object.set_temporary (self.mcp)
   create_start_points (self)
end

local function stop (self)
   if self.handle and self.timeSlice then self.timeSlice:destroy (self.handle) end
end

function new (config, name)
   local self = {
      start_plugin = start,
      stop_plugin = stop,
      log = dmz.log.new ("lua." .. name),
      timeSlice = dmz.time_slice.new (),
      objObs = dmz.object_observer.new (),
      config = config,
      name = name,
      cycleList = {},
      cycleCount = 0,
      startPoints = {},
      maxCycles = config:to_number ("cycles.max", 6),
      minCycles = config:to_number ("cycles.min", 2),
   }

   self.log:info ("Creating plugin: " .. name)
   
   return self
end

File: scripts/mcp.js
var dmz =
      { object: require("dmz/components/object")
      , time: require("dmz/runtime/time")
      , consts: require("const")
      , input: require("dmz/components/input")
      , portal: require("dmz/components/portal")
      , vector: require("dmz/types/vector")
      , defs: require("dmz/runtime/definitions")
      , matrix: require("dmz/types/matrix")
      , util: require("dmz/types/util")
      }

   , CycleList = {}
   , CycleCount = 0
   , StartPoints = []
   , WaitTime = dmz.time.getFrameTime() + 1
   , MaxCycles = self.config.number("cycles.max", 6)
   , MinCycles = self.config.number("cycles.min", 2)
   , MCP = dmz.object.create(dmz.consts.MCPType)
   , AssignPoints = false
   , timeSlice
   , PlayerCount
   , NextStateTime
   , EndTime

   , createStartPoints
   , assignPointsToCycles
   , isAllDrones
   , createStartPoints
   , assignPointsToCycles
   , getStandbyCount
   , getDeadCount
   , setGameState
   , updateTimeSlice
   ;

createStartPoints = function () {
   var Distance = 10
     , Space = 5
     , Offset = Math.floor(MaxCycles / 4) + 1
     , Ident = dmz.matrix.create()
     , HeadingPi = dmz.matrix.create().fromAxisAndAngle(dmz.vector.Up, Math.PI)
     , oddPos = [ -(Space * Offset), 0, Distance ]
     , evenPos = [ -(Space * Offset) + (Space / 2), 0, -Distance ]
     , Odd
     , point
     , ix
     ;

   for (ix = 0; ix < MaxCycles; ix += 1) {
      Odd = (1 == (ix % 2));
      point = {
         index: ix
         , pos: (Odd ? oddPos : evenPos)
         , ori: (Odd ? Ident : HeadingPi)
         , handle: dmz.object.create(dmz.consts.StartPointType)
      };
      StartPoints[ix] = point;
      dmz.object.position(point.handle, null, point.pos);
      dmz.object.orientation(point.handle, null, point.ori);
      // Counter is used to trigger a network send
      dmz.object.counter(point.handle, dmz.consts.StartLinkHandle, 0);
      dmz.object.counter.rollover(point.handle, dmz.consts.StartLinkHandle, true);
      dmz.object.activate(point.handle);
      if (Odd) { oddPos[0] += Space; }
      else { evenPos[0] += Space; }
   }
};

assignPointsToCycles = function () {
   var count = 0
     , cycle
     , setPoint
     ;

   setPoint = function (cycle, count) {
      cycle.point = StartPoints[count];
      dmz.object.link(
         dmz.consts.StartLinkHandle,
         cycle.point.handle,
         cycle.handle);
      // Adding to the objs counter will cause the network rules to
      // send the packet immediately instead of waiting for the one second
      // heartbeat as specified in the net rules for this obj type.
      dmz.object.addToCounter(cycle.point.handle, dmz.consts.StartLinkHandle);
   }

   PlayerCount = 0;
   Object.keys(CycleList).forEach(function (key) {
      cycle = CycleList[key];
      if (cycle.point) {
         dmz.object.unlinkSubObjects(cycle.point.handle, dmz.consts.StartLinkHandle)
         delete cycle.point;
      }
   });

   Object.keys(CycleList).forEach(function (key) {
      cycle = CycleList[key];
      if (count < MaxCycles) {
         if (!cycle.drone) {
            setPoint (cycle, count);
            PlayerCount += 1;
            count += 1;
         }
      }
   });

   Object.keys(CycleList).forEach(function (key) {
      cycle = CycleList[key];
      if (count < MaxCycles) {
         if (cycle.drone) {
            setPoint (cycle, count);
            PlayerCount += 1;
            count += 1;
         }
      }
   });
};

getStandbyCount = function () {
   var result = 0
     , cycle
     ;

   Object.keys(CycleList).forEach(function (key) {
      cycle = CycleList[key];
      if (cycle.point && cycle.standby) { result += 1; }
   });
   return result;
};

getDeadCount = function () {
   var result = 0
     , cycle
     ;

   Object.keys(CycleList).forEach(function (key) {
      cycle = CycleList[key];
      if (cycle.point && cycle.dead) { result += 1; }
   });

   return result;
};

isAllDrones = (self.config.string("drone-free-play.value", "false") === "true") ?
function () { return false; } :
function () {
   var result = true
     , cycle
     ;

   Object.keys(CycleList).forEach(function (key) {
      cycle = CycleList[key];
      if (cycle.point && !cycle.dead && !cycle.drone) { result = false; }
   });
   return result;
};

setGameState = function (mcp, State) {
   var newState = dmz.object.state(mcp);
   newState = newState.unset(dmz.consts.GameStateMask);
   newState = newState.or(State);
   dmz.object.state(mcp, null, newState);
};

updateTimeSlice = function (time) {

   var State = dmz.object.state(MCP)
     , CTime
     ;

   if (State) {
      CTime = dmz.time.getFrameTime();
      if (State.contains(dmz.consts.GameWaiting)) {
         if (AssignPoints) {
            assignPointsToCycles();
            AssignPoints = false;
            WaitTime = CTime + 1;
         }
         if ((WaitTime <= CTime) &&
               (getStandbyCount() == PlayerCount) &&
               (PlayerCount >= MinCycles)) {
            setGameState(MCP, dmz.consts.GameCountdown5);
            NextStateTime = CTime + 1;
         }
      }
      else if (State.contains(dmz.consts.GameActive)) {
         if (EndTime && (EndTime <= CTime)) {
            setGameState(MCP, dmz.consts.GameWaiting);
            EndTime = null;
            WaitTime = CTime + 2;
         }
         else if (!EndTime && (getDeadCount() >= (PlayerCount - 1)
                 || isAllDrones())) {
            EndTime = CTime + 2;
         }
      }
      else if (State.contains(dmz.consts.GameCountdown5)) {
         if (CTime >= NextStateTime) {
            setGameState(MCP, dmz.consts.GameCountdown4);
            NextStateTime = CTime + 1;
         }
      }
      else if (State.contains(dmz.consts.GameCountdown4)) {
         if (CTime >= NextStateTime) {
            setGameState(MCP, dmz.consts.GameCountdown3);
            NextStateTime = CTime + 1;
         }
      }
      else if (State.contains(dmz.consts.GameCountdown3)) {
         if (CTime >= NextStateTime) {
            setGameState(MCP, dmz.consts.GameCountdown2);
            NextStateTime = CTime + 1;
         }
      }
      else if (State.contains(dmz.consts.GameCountdown2)) {
         if (CTime >= NextStateTime) {
            setGameState(MCP, dmz.consts.GameCountdown1);
            NextStateTime = CTime + 1;
         }
      }
      else if (State.contains(dmz.consts.GameCountdown1)) {
         if (CTime >= NextStateTime) {
            setGameState(MCP, dmz.consts.GameActive);
            NextStateTime = CTime + 1;
         }
      }
      else {
         self.log.error("Game in unknown state. Changing to a waiting state");
         setGameState(MCP, dmz.consts.GameWaiting);
      }
   }
};



dmz.object.create.observe(self, function (obj, Type) {
   if (Type.isOfType(dmz.consts.CycleType)) {
      CycleList[obj] =
         { handle: obj
         , drone: dmz.object.flag(obj, dmz.consts.DroneHandle)
         };
      CycleCount += 1;
      AssignPoints = true;
   }
});

dmz.object.destroy.observe(self, function (obj, Type) {
   var cycle = CycleList[obj];
   if (cycle) {
      delete CycleList[obj];
      CycleCount -= 1;
      if (cycle.point) { PlayerCount -= 1; }
      AssignPoints = true;
   }
});

dmz.object.state.observe(self, function (obj, Attribute, State, PreviousState) {
   var cycle = CycleList[obj];
   if (cycle) {
      if (!PreviousState) { PreviousState = dmz.consts.EmptyState; }
      if (State.contains(dmz.consts.Dead) &&
            !PreviousState.contains(dmz.consts.Dead)) {
         cycle.dead = true;
         cycle.standby = false;
      }
      else if (State.contains(dmz.consts.Standby)
            && !PreviousState.contains(dmz.consts.Standby)) {
         cycle.dead = false;
         cycle.standby = true;
      }
      else {
         cycle.dead = false;
         cycle.standby = false;
      }
   }
});

(function () {
   dmz.object.state(MCP, null, dmz.consts.GameWaiting);
   dmz.object.activate(MCP);
   createStartPoints();
}());

timeSlice = dmz.time.setRepeatingTimer(self, updateTimeSlice);
